aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -395,26 +448,39 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
   int64_t nDim = dense.dim();
   int64_t nDimI = sparse.sparse_dim();
 
-  if (!is_same_tensor(r, dense)) r.copy_(dense);
-  if (sparse._nnz() == 0) return r;
+  if (sparse._nnz() == 0) {
+    if (!is_same_tensor(r, dense)) r.copy_(dense);
+    return r;
+  }
+
+  Tensor valuesBuffer = values.to(commonDtype);
+  Tensor resultBuffer = r;
+  if (r.scalar_type() != commonDtype) {
+    resultBuffer = dense.to(commonDtype);
+  } else if (!is_same_tensor(r, dense)) {
+    resultBuffer.copy_(dense);
+  }
 
   // accessors rely on nnz test
   if (nDim > nDimI) {
     auto indices_accessor = indices.accessor<int64_t, 2>();
     for (int64_t k = 0; k < sparse._nnz(); k++) {
-      Tensor dstBuffer = r;
+      Tensor& dstBuffer = resultBuffer;
```

Comment:
@zou3519 This suggested change from Tensor to Tensor& wasn't in the original code, and seems to have caused unit tests to fail. I'm going to change this back.

Reply:
What’s wrong with it? We shouldn’t revert lines just because the unit tests fail; we should understand what the problem is.

Reply:
This is just refraining from making an additional change to the underlying implementation of add. There are infinite possible changes we could make, this one didn't work, and if we did make it doesn't actually seem related to this PR either.

Reply:
It turns out that my original suggestion doesn't work because later on we're assigning a Tensor to dstBuffer:
```
for (int64_t d = 0; d < sparse.sparse_dim(); d++) {
    dstBuffer = dstBuffer.select(0, indices_accessor[d][k]);
}
```
Assigning the rvalue Tensor to a Tensor& is wrong because the Tensor gets cleaned up instantly, leading to the reference dstBuffer pointing to garbage. Changing it back here is correct.